perun-core/src/main/java/cz/metacentrum/perun/core/impl/Utils.java

```diff
@@ -1009,84 +1007,106 @@ public static String getMessageAuthenticationCode(String input) {
+	 *
+	 * @param user user to send notification for
+	 * @param email user's email to send notification to
+	 * @param namespace namespace to reset password in
+	 * @param url base URL of Perun instance
+	 * @param id ID of account activation request
+	 * @param messageTemplate message of the email (use default if null)
+	 * @param subject subject of the email (use default if null)
+	 */
+	public static void sendAccountActivationEmail(User user, String email, String namespace, String url, int id, String messageTemplate, String subject) {
+		String instanceName = BeansUtils.getCoreConfig().getInstanceName();
 
-		} catch (UnsupportedEncodingException ex) {
-			throw new InternalErrorException("Unable to encode validation URL for mail change.", ex);
-		} catch (MalformedURLException ex) {
-			throw new InternalErrorException("Not valid URL of running Perun instance.", ex);
+		String validationLink = prepareValidationLinkForPasswordResetAndAccountActivation(url, "/non/pwd-reset/", id, user, namespace);
+		String validityTo = prepareValidityTo();
+
+		String defaultSubject = "[" + instanceName + "] Account activation in namespace: " + namespace;
+		String defaultBody = "Dear " + user.getDisplayName() + ",\n\nWe've received request to activate your account in namespace \"" + namespace + "\"." +
+			"\n\nPlease visit the link below, where you can activate your account:\n\n" + validationLink + "\n\n" +
+			"Link is valid till " + validityTo + "\n\n" +
+			"Message is automatically generated." +
+			"\n----------------------------------------------------------------" +
+			"\nPerun - Identity & Access Management System";
+
+		Map<String, String> subjectParametersToReplace = new HashMap<>();
+		subjectParametersToReplace.put("{instanceName}", instanceName);
+		subjectParametersToReplace.put("{namespace}", namespace);
+		subject = prepareSubjectOfEmail(subject, defaultSubject, subjectParametersToReplace);
+
+		Map<String, String> bodyParametersToReplace = new HashMap<>();
+		bodyParametersToReplace.put("{displayName}", user.getDisplayName());
+		bodyParametersToReplace.put("{namespace}", namespace);
+		bodyParametersToReplace.put("{validity}", validityTo);
+		// allow enforcing per-language links
+		if (messageTemplate.contains("{link-")) {
+			Pattern pattern = Pattern.compile("\\{link-[^}]+}");
+			Matcher matcher = pattern.matcher(messageTemplate);
+			while (matcher.find()) {
+				// whole "{link-something}"
+				String toSubstitute = matcher.group(0);
+				String langLink = validationLink;
+
+				Pattern namespacePattern = Pattern.compile("-(.*?)}");
+				Matcher m2 = namespacePattern.matcher(toSubstitute);
+				if (m2.find()) {
+					// only language "cs", "en",...
+					String lang = m2.group(1);
+					langLink = langLink + "&locale=" + lang;
+				}
+				bodyParametersToReplace.put(toSubstitute, langLink);
+			}
+		} else {
+			bodyParametersToReplace.put("{link}", validationLink);
```

Comment:
We must append ``&activation=true`` so that app for password reset knows, it should behave as an account activation. Or you can handle this in link preparation method.

Reply:
Added to preparation of link.